### PR TITLE
Update 2d fdsp all gather 

### PR DESCRIPTION
**Pull request by suexu1025**
# Description
* update 2d fdsp all gather with consideriation of gather fsdp on expert dim first to get rid of extra data format
* enable `attn_logits_soft_cap` as option for benchmark only
# Tests
http://xprof/?session_id=amandaliang-5205770298363717970 
Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/layers/moe.py [hunk 1]
```diff
@@ -1256,11 +1256,11 @@ def get_active_sharding_axes(pspec_dim_axes, tensor_dim_index):
 
     if self.config.moe_fsdp_use_two_stage_all_gather:
       # Unshard on fsdp axis
-      w0_kernel = self._maybe_shard_with_logical(w0_kernel, ("exp", "embed_tensor_transpose", "mlp"))
-      w1_kernel = self._maybe_shard_with_logical(w1_kernel, ("exp", "embed_tensor_transpose", "mlp"))
+      w0_kernel = nn.with_logical_constraint(w0_kernel, ("exp_with_fsdp", "embed_tensor_transpose", "mlp"))
+      w1_kernel = nn.with_logical_constraint(w1_kernel, ("exp_with_fsdp", "embed_tensor_transpose", "mlp"))
```
Comment (NuojCheng) [inline src/MaxText/layers/moe.py]: Suggest using `self._maybe_shard_with_logical`

File: src/MaxText/layers/moe.py [hunk 2]
```diff
@@ -1269,9 +1269,9 @@ def get_active_sharding_axes(pspec_dim_axes, tensor_dim_index):
       wo_kernel = jax.lax.optimization_barrier(wo_kernel)
 
       # Unshard on both fsdp and fsdp_transpose transpose
-      w0_kernel = self._maybe_shard_with_logical(w0_kernel, ("exp", "embed_tensor_transpose", "mlp_no_fsdp"))
-      w1_kernel = self._maybe_shard_with_logical(w1_kernel, ("exp", "embed_tensor_transpose", "mlp_no_fsdp"))
-      wo_kernel = self._maybe_shard_with_logical(wo_kernel, ("exp", "mlp_no_fsdp", "embed_tensor_transpose"))
+      w0_kernel = self.with_logical_constraint(w0_kernel, ("exp_with_fsdp", "embed_tensor_transpose", "mlp_no_fsdp"))
+      w1_kernel = self.with_logical_constraint(w1_kernel, ("exp_with_fsdp", "embed_tensor_transpose", "mlp_no_fsdp"))
+      wo_kernel = self.with_logical_constraint(wo_kernel, ("exp_with_fsdp", "mlp_no_fsdp", "embed_tensor_transpose"))
```
Comment (BirdsOfAFthr) [inline src/MaxText/layers/moe.py]: `nn.with_logical_constraint`

**Review by BirdsOfAFthr — APPROVED** [review]

File: src/MaxText/layers/moe.py (hunk 2 above)
Comment (suexu1025) [reply src/MaxText/layers/moe.py]: updated with self._maybe_shard_with_logical, I verified the perf which is aligned with nn. with_logical_constraint

File: src/MaxText/layers/moe.py (hunk 1 above)
Comment (suexu1025) [reply src/MaxText/layers/moe.py]: updated

**Review by NuojCheng — APPROVED** [review]
LGTM!
